src/operations.cpp

```diff
@@ -614,8 +613,51 @@ namespace
   
   typedef bool (*Ptr_equal_string_ordinal_ic)(const wchar_t*, const wchar_t*);
 
-  Ptr_equal_string_ordinal_ic equal_string_ordinal_ic = 
-    rtl_equal_unicode_string_api ? equal_string_ordinal_ic_1 : equal_string_ordinal_ic_2;
+  BOOST_NOINLINE Ptr_equal_string_ordinal_ic get_compare_function()
+  {
+    rtl_equal_unicode_string_api = PtrRtlEqualUnicodeString(
+      ::GetProcAddress(
+        ::GetModuleHandleW(L"ntdll.dll"), "RtlEqualUnicodeString"));
+
+    return rtl_equal_unicode_string_api ? equal_string_ordinal_ic_1
+                                        : equal_string_ordinal_ic_2;
+  }
+
+  static Ptr_equal_string_ordinal_ic compare_function;
+  static long compare_function_state;
+
+  inline Ptr_equal_string_ordinal_ic compare()
+  {
+    for (unsigned i = 0; compare_function_state < 2; ++i) 
```

Comment:
This reads `compare_function_state` non-atomically (i.e. not thread-safe).

Reply:
Ouch, you are completely right! I thought that I could use the `<atomic>`-free interlocked approach that `boost/interprocess/detail/win32_api.hpp` uses, but now I realize that the corresponding code in the `function_address_holder::get_module` and `function_address_holder::get` is defect as well! I will take a look this night or tomorrow to improve the patch - thanks for the review so far!

Reply:
The non-atomical read access should have been fixed now via https://github.com/Dani-Hub/filesystem/commit/76c061ec79328cc4fee6a23f07e2918c66fe44a7

Reply:
I have now created ticket https://svn.boost.org/trac10/ticket/13309 to handle the same problem in `boost/interprocess/detail/win32_api.hpp` (twice)